flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorageForRecovery.java

```diff
@@ -33,8 +34,29 @@
 public class FsStateChangelogStorageForRecovery
         implements StateChangelogStorageView<ChangelogStateHandleStreamImpl> {
 
+    private final ChangelogStreamHandleReaderWithCache changelogHandleReaderWithCache;
+
+    public FsStateChangelogStorageForRecovery() {
+        this.changelogHandleReaderWithCache = null;
+    }
+
+    public FsStateChangelogStorageForRecovery(Configuration configuration) {
+        this.changelogHandleReaderWithCache =
+                new ChangelogStreamHandleReaderWithCache(configuration);
+    }
+
     @Override
     public StateChangelogHandleReader<ChangelogStateHandleStreamImpl> createReader() {
-        return new StateChangelogHandleStreamHandleReader(new StateChangeFormat());
+        return new StateChangelogHandleStreamHandleReader(
+                changelogHandleReaderWithCache != null
+                        ? new StateChangeIteratorImpl(changelogHandleReaderWithCache)
+                        : new StateChangeIteratorImpl());
```

Comment:
Can't the logic of choosing `ChangelogStreamHandleReader` be moved to the creator of this object?

So that:
- `FsStateChangelogStorage` passes no-caching version to `FsStateChangelogStorageForRecovery` constructor
- `FsStateChangelogStorageFactory` passes configuration, from which a caching version is built

(so there is no `null` and no `if` here)

Reply:
Thanks @rkhachatryan !